File: rust/sum_of_multiples.rs
use std::collections::HashSet;

pub fn sum_of_multiples(limit: u32, factors: &[u32]) -> u32 {
    let mut set = HashSet::new();
    for &f in factors {
        if f == 0 {
            continue;
        }
        let mut g = f;
        while g < limit {
            set.insert(g);
            g += f;
        }
    }

    set.into_iter().sum()
}

File: zig/sum_of_multiples.zig
const std = @import("std");
const mem = std.mem;

pub fn sum(allocator: mem.Allocator, factors: []const u32, limit: u32) !u64 {
    var mp = std.AutoHashMap(u32, void).init(allocator);
    defer mp.deinit();
    var ans: u64 = 0;
    for (factors) |f| {
        if (f == 0) {
            continue;
        }
        var n = f;
        while (n < limit) : (n += f) {
            const v = try mp.getOrPut(n);
            if (!v.found_existing) {
                ans += n;
            }
        }
    }
    return ans;
}
